phenopacketlab-spring-boot-autoconfigure/src/main/java/org/monarchinitiative/phenopacketlab/autoconfigure/ConceptResourceServiceLoader.java

```diff
@@ -0,0 +1,151 @@
+import org.monarchinitiative.phenopacketlab.io.HgncConceptLoader;
+import org.monarchinitiative.phenopacketlab.model.IdentifiedConceptResource;
+import org.monarchinitiative.phenopacketlab.core.ConceptResourceServiceImpl;
+import org.monarchinitiative.phenopacketlab.model.OntologyConceptResource;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Class for parallel loading of {@link ConceptResourceService}.
+ */
+class ConceptResourceServiceLoader {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConceptResourceServiceLoader.class);
+
+    private final ExecutorService executor;
+    private final PhenopacketLabDataResolver dataResolver;
+
+    ConceptResourceServiceLoader(ExecutorService executor,
+                                 PhenopacketLabDataResolver dataResolver) {
+        this.executor = Objects.requireNonNull(executor);
+        this.dataResolver = Objects.requireNonNull(dataResolver);
+    }
+
+    ConceptResourceService load() throws InvalidResourceException {
+        Resources result = new Resources();
+        List<String> errors = Collections.synchronizedList(new LinkedList<>());
+        List<ResourceTuple<?>> resources = List.of(
+                new ResourceTuple<>(dataResolver.efoJsonPath(), ConceptResourceLoaders::efo, result::setEfo),
+                new ResourceTuple<>(dataResolver.genoJsonPath(), ConceptResourceLoaders::geno, result::setGeno),
+                new ResourceTuple<>(dataResolver.hgncCompleteSetPath(), is -> HgncConceptLoader.load(is, "HGNC_VERSION"), result::setHgnc),
+                new ResourceTuple<>(dataResolver.hpJsonPath(), ConceptResourceLoaders::hpo, result::setHp),
+                new ResourceTuple<>(dataResolver.mondoJsonPath(), ConceptResourceLoaders::mondo, result::setMondo),
+                new ResourceTuple<>(dataResolver.soJsonPath(), ConceptResourceLoaders::so, result::setSo),
+                new ResourceTuple<>(dataResolver.uberonJsonPath(), ConceptResourceLoaders::uberon, result::setUberon),
+                new ResourceTuple<>(dataResolver.ncitJsonPath(), ConceptResourceLoaders::ncit, result::setNcit),
+                new ResourceTuple<>(dataResolver.gssoJsonPath(), ConceptResourceLoaders::gsso, result::setGsso)
+        );
+
+        CountDownLatch latch = new CountDownLatch(resources.size());
+        for (ResourceTuple<?> resource : resources) {
+            executor.submit(prepareTask(resource, errors::add, latch));
+        }
+
+        try {
+            latch.await();
```

Comment:
Do we want to await forever or would the await(time, timeunit) be better?

Reply:
I increased the verbosity of the parallel loading. Now the main thread checks each 20 seconds if the loading has been finished and logs a statement into the console. However, the only effect of the changes is that the app will indicate that something is not right. Typically, the loading should take less. The backend has no use if loading of a resource fails.

On a side note, the current implementation is likely only a temporary, to allow the development. In real-life, phenopackets are free to use *any* release of *any* [Resource](https://phenopacket-schema.readthedocs.io/en/latest/resource.html). As a consequence, we may need to manage multiple versions/releases of NCIT, HPO, MONDO, and other controlled vocabularies in the backend. In that case, the simple setup where we keep the resource files in a single data directory (`phenopacketlab.data-directory` at `phenopacketlab-restapi/src/main/resources/application-development.yml`) will not work. There are multiple ways how we can do this, we'll need to work on that down the road. I think it becomes relevant in Milestone [here](https://github.com/TheJacksonLaboratory/PhenopacketLab/milestone/5).